### DISCUSSION TITLE
Flux diff with Flux Kustomization

### DISCUSSION THREAD

**Discussion by vterdunov** [fluxcd/flux2#2954] [Q&A] [answered]
Examples says:
```yaml
# Preview using a local flux kustomization file
flux diff kustomization my-app --path ./path/to/local/manifests --kustomization-file ./path/to/local/my-app.yaml
```

1) It's a bit implicit to me. Which information Flux CLI get from the Flux Kustomization?
`spec.path` already defined or implicitly translates to the root path of the SourceRef. Why we have to pass `--path` explicitly then?

What happened if pass different pathes to `--path` and `spec.path`?

2) Flux Kustomization has `metadata.name`. Why we have to pass kustomization name then? Maybe it required if a file contains more than one Flux Kustomization.

**Comment by kingdonb** [comment, marked as answer]
When you run `flux diff kustomization my-app`, the Flux Kustomization needs to know where on the local filesystem to locate the path that Kustomization targets. Flux does not assume that you run `flux diff kustomization` from the root path of the SourceRef. There might be opportunities for integrations like the [VS Code GitOps Tools](https://marketplace.visualstudio.com/items?itemName=Weaveworks.vscode-gitops-tools) to obviate this requirement, when the commands are known to be executed at the root of the SourceRef.

Thank you for providing this feedback, it's helpful to understand how people want to use Flux.

> Flux Kustomization has `metadata.name`. Why we have to pass kustomization name then?

Same answer. You may have a point about this generally, it should be possible to infer the name and namespace from the Flux Kustomization file on disk. But `--kustomization-file` is not a mandatory option, it's only needed to be passed when you have made changes to the Flux Kustomization file that affect the result, like for example, changing the value of `spec.prune` from false to true.

If you do not provide a `kustomization-file` then the version in the cluster will be used without modification. If you do provide a kustomization-file, then it makes sense we could usually infer the name and namespace from its contents. (This might not be 100% reliable though, because there could always be another kustomize CLI overlay between that kustomize-file and where it gets applied.)

```
~/projects/csh-flux/secrets/home-workers $ flux diff kustomization my-secrets --path ./ --kustomization-file ../../clusters/home-workers/my-secrets/my-secrets-kustomization.yaml
✓  Kustomization diffing...
► Secret/flux-system/webhook-token created
⚠️ identified at least one change, exiting with non-zero exit code
```

Here's an example invocation that should help clarify why you want to be able to set `--path` – I'm not in the repo root, I'm in the directory where it's applied from. It would be awkward if the only way to invoke `flux diff` from here was to run `cd ../../` first.
